syncer/syncer_test.go

```diff
@@ -80,11 +101,74 @@ func (s *testSyncerSuite) SetUpSuite(c *C) {
 
 	s.resetMaster()
 	s.resetBinlogSyncer()
+	s.resetEventsGenerator(c)
 
 	_, err = s.db.Exec("SET GLOBAL binlog_format = 'ROW';")
 	c.Assert(err, IsNil)
 }
 
+func (s *testSyncerSuite) generateEvents(binlogEvents mockBinlogEvents, c *C) []*replication.BinlogEvent {
+	events := make([]*replication.BinlogEvent, 0, 1024)
+	for _, e := range binlogEvents {
+		switch e.typ {
+		case DBCreate:
+			evs, _, err := s.eventsGenerator.GenCreateDatabaseEvents(e.args[0].(string))
+			c.Assert(err, IsNil)
+			events = append(events, evs...)
+		case DBDrop:
+			evs, _, err := s.eventsGenerator.GenDropDatabaseEvents(e.args[0].(string))
+			c.Assert(err, IsNil)
+			events = append(events, evs...)
+		case TableCreate:
+			evs, _, err := s.eventsGenerator.GenCreateTableEvents(e.args[0].(string), e.args[1].(string))
+			c.Assert(err, IsNil)
+			events = append(events, evs...)
+		case TableDrop:
+			evs, _, err := s.eventsGenerator.GenDropTableEvents(e.args[0].(string), e.args[1].(string))
+			c.Assert(err, IsNil)
+			events = append(events, evs...)
+
+		case Write, Update, Delete:
+			dmlData := []*event.DMLData{
+				{
+					TableID:    e.args[0].(uint64),
+					Schema:     e.args[1].(string),
+					Table:      e.args[2].(string),
+					ColumnType: e.args[3].([]byte),
+					Rows:       e.args[4].([][]interface{}),
+				},
+			}
+			eventType := replication.WRITE_ROWS_EVENTv2
+			if e.typ == Update {
+				eventType = replication.UPDATE_ROWS_EVENTv2
+			} else if e.typ == Delete {
+				eventType = replication.DELETE_ROWS_EVENTv2
+			}
+			evs, _, err := s.eventsGenerator.GenDMLEvents(eventType, dmlData)
+			c.Assert(err, IsNil)
+			events = append(events, evs...)
+		}
+
+	}
+	return events
+}
+
+func (s *testSyncerSuite) resetEventsGenerator(c *C) {
+	previousGTIDSetStr := "3ccc475b-2343-11e7-be21-6c0b84d59f30:1-14,406a3f61-690d-11e7-87c5-6c92bf46f384:1-94321383"
+	previousGTIDSet, err := gtid.ParserGTID(s.cfg.Flavor, previousGTIDSetStr)
+	if err != nil {
+		c.Fatal("", zap.Error(err))
+		c.FailNow()
+	}
+	latestGTIDStr := "3ccc475b-2343-11e7-be21-6c0b84d59f30:14"
+	latestGTID, err := gtid.ParserGTID(s.cfg.Flavor, latestGTIDStr)
+	s.eventsGenerator, err = event.NewGenerator(s.cfg.Flavor, uint32(s.cfg.ServerID), 0, latestGTID, previousGTIDSet, 0)
+	if err != nil {
+		c.Fatal("", zap.Error(err))
+		c.FailNow()
```

Comment:
already run c.Fatal before, is this FailNow worked?

Reply:
Fixed